### ISSUE TITLE
Feature proposal: Multi-target DMatrices and Boosters for C API

### ISSUE THREAD

**Issue by jmaerte** [dmlc/xgboost#10714]
**Problem**: Multi-target Regression on external memory (e.g. following https://github.com/dmlc/xgboost/blob/master/demo/c-api/external-memory/external_memory.c#L21C1-L22C18) seems to not be possible without iterating through the entirety of the memory as many times as you have target variables.

**Possible solution**: Either allow DMatrix/ProxyDMatrix to hold multiply target values or even a target-matrix (with a similar JSON string encoding array interface (see line 112 in above file) or allow for the external memory module to compress multiple y-arrays (1d) together with the X-Matrices so that everything can be loaded from the temporary compression file (i.e. from the compression there can be made k DMatrices each with the same X but different y values for training k different models).

**Oddities**: From a runtime-perspective it seems to me that the xgboost python package already allows for exactly this behavior; I do not necessarily need Multi-Target boosters as it is One-Model-Per-Target anyway but building the model from external memory each time is very time consuming and seems unnecessary; python does seem to be able to learn a multi-target regressor from external memory in one iteration, however.

If this feature request (though it seems to be solved already in the python wrapper code) is not to be implemented in the original C API maybe you do have some instructions or tips & tricks on how to smuggle multiple y values into the DMatrix?

I am looking forward to your response!

**Comment by trivialfis** [comment]
You can do that with C as well, but internally XGBoost still iterate through the data multiple times during model training. We are working on the external memory support for vector leaf, which uses a single model, stay tuned.

**Comment by jmaerte** [comment]
Thank you for the quick response.

I don‘t mind XGBoost internally iterating over the y‘s as long as i do not need to compress the entire external memory k times.

Could you maybe provide a quick code snipped for how to set the y of a ProxyDMatrix to be multi-column? Maybe via the json string interface? That would already be very helpful for me, as currently when setting the y pointer to a multi-column array and setting an interface string analogous to the one of X XGBoost learns the target values as if y is single column.

Best regards,
Julian Märte

**Comment by trivialfis** [comment]
https://github.com/dmlc/xgboost/blob/fd365c147ef36fcc4ea5caa3a5625842e2515a01/src/c_api/c_api.cc#L617
Should do the job. If you are looking into Python, then it's a good start. The native Python package is built on top of C with very little auxiliary code. Most of the Python code are for integration with other frameworks like sklearn

**Comment by jmaerte** [comment]
Even though I had already tried this function for char const* field == "target" and a numpy json string equivalent to the one I've given for X without any success (it just learned it as if the given target pointer represented a 1d array) I will close this for now and guess I'll wait until there is some more material from you once you launch the multi-target regression for vector leaf...

**Comment by trivialfis** [comment]
The field is label instead of target
